src/pysdmx/model/__base.py

```diff
@@ -0,0 +1,212 @@
+    """Annotable Artefact class.
+
+    Superclass of all SDMX artefacts.
+    Contains the list of annotations.
+
+    Attributes:
+        annotations: The list of annotations attached to the artefact.
+    """
+
+    annotations: Sequence[Annotation] = ()
+
+    @classmethod
+    def __all_annotations(cls) -> Dict[str, Any]:
+        class_attributes = {}
+        for c in cls.__mro__:
+            if "__annotations__" in c.__dict__:
+                class_attributes.update(c.__annotations__)
+        return dict(reversed(list(class_attributes.items())))
+
+    def __str__(self) -> str:
+        """Returns a human-friendly description."""
+        out = []
+        for k in self.__all_annotations().keys():
+            v = self.__getattribute__(k)
+            if v:
+                out.append(f"{k}={str(v)}")
+        return ", ".join(out)
+
+
+class IdentifiableArtefact(AnnotableArtefact, frozen=True, omit_defaults=True):
+    """Identifiable Artefact class.
+
+    Provides identity to all derived classes. It also provides annotations
+    to derive classes because it is a subclass of AnnotableArtefact.
+
+    Attributes:
+        id: The identifier of the artefact.
+        uri: The URI of the artefact.
+        urn: The URN of the artefact.
+    """
+
+    id: str = ""
+    uri: Optional[str] = None
+    urn: Optional[str] = None
+
+
+class NameableArtefact(IdentifiableArtefact, frozen=True, omit_defaults=True):
+    """Nameable Artefact class.
+
+    Provides a name and a description to all derived classes.
+
+    Attributes:
+        name: The name of the artefact.
+        description: The description of the artefact.
+    """
+
+    name: Optional[str] = None
+    description: Optional[str] = None
+
+
+class VersionableArtefact(NameableArtefact, frozen=True, omit_defaults=True):
+    """Versionable Artefact class.
+
+    Provides a version to all derived classes.
+
+    Attributes:
+        version: The version of the artefact.
+        valid_from: The date from which the artefact is valid.
+        valid_to: The date to which the artefact is valid.
+    """
+
+    version: str = "1.0"
+    valid_from: Optional[datetime] = None
+    valid_to: Optional[datetime] = None
+
+
+class MaintainableArtefact(
+    VersionableArtefact, frozen=True, omit_defaults=True
+):
+    """Maintainable Artefact class.
+
+    An abstract class to group together primary structural metadata
+    artefacts that are maintained by an Agency.
+
+    Attributes:
+        is_final: Whether the artefact is final.
+        is_external_reference: Whether the artefact is an external reference.
+        service_url: The URL of the service.
+        structure_url: The URL of the structure.
+        agency: The maintainer of the artefact.
+    """
+
+    is_final: bool = False
+    is_external_reference: bool = False
+    service_url: Optional[str] = None
+    structure_url: Optional[str] = None
+    agency: Optional[Union[str, "Agency"]] = None
```

Comment:
I think agency should not be optional for maintainable artefacts?

Reply:
If we move Contact, Item, Organisation, Agency, etc above, I guess we don't need to use the quotes here?

Reply:
Agree on Agency should not be optional, what should it be the default value for Agency? (None?, ""?). 

After some reorganisation I moved them to prevent the use of quotes in this file.

Reply:
Removed quotes on https://github.com/bis-med-it/pysdmx/pull/27/commits/40e23ff46b381c0287356ee437eac75fe92eedb0

Reply:
Okay, let´s go for it

Reply:
src\pysdmx\model\__base.py:197:34: error: Incompatible types in assignment (expression has type "None", variable has type "str | Agency")  [assignment]

Cannot have None as default value if we remove the Optional typing. Should we ignore it? @sosna

Reply:
> Cannot have None as default value if we remove the Optional typing. Should we ignore it? @sosna 

You are right. Maybe we should default to an empty string then, and, as discussed, check in the post init function that a value has been set?

Reply:
Yes, adding a new issue with this.

Reply:
Mentioned in #34